app/commons/notification_service/client.py

```diff
@@ -0,0 +1,103 @@
+# Copyright (C) 2022 Indoc Research
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from typing import List, Optional
+from uuid import UUID
+
+import httpx
+from common import LoggerFactory
+
+from app.commons.meta_services import bulk_get_by_ids
+from app.commons.notification_service.models import (
+    CopyRequestAction,
+    Location,
+    Node,
+    NotificationType,
+    PipelineNotification,
+    Target,
+    TargetType,
+)
+from app.config import ConfigClass
+
+logger = LoggerFactory(
+    'api_copy_request',
+    level_default=ConfigClass.LOG_LEVEL_DEFAULT,
+    level_file=ConfigClass.LOG_LEVEL_FILE,
+    level_stdout=ConfigClass.LOG_LEVEL_STDOUT,
+    level_stderr=ConfigClass.LOG_LEVEL_STDERR,
+).get_logger()
+
+
+class NotificationServiceClient:
+    """Client for sending notifications into Notification Service."""
+
+    def __init__(
+        self,
+        recipient_username: str,
+        include_ids: Optional[list[str]],
+        initiator_username: str,
+        source_id: Optional[str],
+        destination_id: Optional[str],
+        project_code: str,
+        action: CopyRequestAction,
+        request_id: UUID,
+        notification_type: NotificationType,
+    ) -> None:
+        self.endpoint = f'{ConfigClass.NOTIFICATION_SERVICE}/v1/all/notifications/'
+        self.recipient_username = recipient_username
+        self.include_ids = include_ids
+        self.initiator_username = initiator_username
+        self.source_id = source_id
+        self.destination_id = destination_id
+        self.project_code = project_code
+        self.action = action
+        self.request_id = request_id
+        self.notification_type = notification_type
+
+    async def set_location(self, entity_id: UUID) -> Location:
+        result = await bulk_get_by_ids([entity_id])
+        node = Node(result[0])
+        return Location(id=node.id, path=str(node.display_path), zone=node.zone)
+
+    async def set_targets(self) -> List[Target]:
+        result = await bulk_get_by_ids(self.include_ids)
+        nodes = {node['id']: Node(node) for node in result}
+        targets = []
+        for _node, file_node in nodes.items():
+            targets.append(Target(id=file_node.id, name=file_node.name, type=TargetType(file_node.entity_type)))
+        return targets
+
+    async def send_notifications(self) -> None:
```

Comment:
Based on my previous suggestion this method should accept a pre-created instance of the notification.
```suggestion
    async def send_notification(self, notification: CopyRequestNotification) -> None:
```